include/rmm/mr/device/system_memory_resource.hpp

```diff
@@ -0,0 +1,187 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cstddef>
+#include <limits>
+#include <optional>
+
+namespace rmm::mr {
+
+namespace detail {
+struct sam {
+  static bool is_supported()
+  {
+    int pageableMemoryAccess;
+    RMM_CUDA_TRY(cudaDeviceGetAttribute(&pageableMemoryAccess,
+                                        cudaDevAttrPageableMemoryAccess,
+                                        rmm::get_current_cuda_device().value()));
+    return pageableMemoryAccess == 1;
+  }
+};
+}  // namespace detail
+
+/**
+ * @addtogroup device_memory_resources
+ * @{
+ * @file
+ */
+/**
+ * @brief `device_memory_resource` derived class that uses malloc/free for allocation/deallocation.
+ *
+ * There are two flavors of hardware/software environments that support accessing system allocated
+ * memory from the GPU: HMM and ATS.
+ *
+ * Heterogeneous Memory Management (HMM) is a software-based solution for PCIe-connected GPUs on
+ * x86 systems. Requirements:
+ *   - NVIDIA CUDA 12.2 with the open-source r535_00 driver or newer.
+ *   - A sufficiently recent Linux kernel: 6.1.24+, 6.2.11+, or 6.3+.
+ *   - A GPU with one of the following supported architectures: NVIDIA Turing, NVIDIA Ampere,
+ *     NVIDIA Ada Lovelace, NVIDIA Hopper, or newer.
+ *   - A 64-bit x86 CPU.
+ *
+ *  For more information, see
+ *  https://developer.nvidia.com/blog/simplifying-gpu-application-development-with-heterogeneous-memory-management/.
+ *
+ *  Address Translation Services (ATS) is a hardware/software solution for the Grace Hopper
+ *  Superchip that uses the NVLink Chip-2-Chip (C2C) interconnect to provide coherent memory. For
+ *  more information, see
+ *  https://developer.nvidia.com/blog/nvidia-grace-hopper-superchip-architecture-in-depth/.
+ */
+class system_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new system memory resource.
+   *
+   * By default if no parameters are specified, this memory resource pass through to malloc/free
+   * for allocation/deallocation.
+   *
+   * However, when GPU memory is over-subscribed, system allocated memory will migrate to the GPU
+   * and cannot migrate back, thus causing other CUDA calls to fail with out-of-memory errors. To
+   * work around this problem, we can reserve some GPU memory as headroom for other CUDA calls.
+   * Doing this check can be expensive, so only large buffer above the given threshold will be
+   * checked.
```

Comment:
*nit*: I don't think I quite follow this documentation. The joining `However` conjunction doesn't seem to follow from the previous paragraph.

For example, what is "this check"?

Reply:
This is moved, removed "however".